### PR TITLE
[web] Speed up PageView/CustomPainter rendering

### PR DESCRIPTION
**Pull request by ferhatb**
## Description

Speed up PageView/CustomPainter rendering.
bench_page_view_scroll_line_through.html.drawFrameDuration.average	        2.63x	
bench_page_view_scroll_line_through.html.drawFrameDuration.outlierRatio	1.46x	
bench_page_view_scroll_line_through.html.totalUiFrame.average	                 3.29x

Other benchmarks are in 1.0 to 1.07x range.

## Related Issues

https://github.com/flutter/flutter/issues/65824
https://github.com/flutter/flutter/issues/62596

## Tests

I added the following tests:

page_test.dart

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Reviewer Checklist

- [x] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

### CODE REVIEW

File: lib/web_ui/lib/src/engine/html/path/path.dart [hunk 1]
```diff
@@ -1537,6 +1537,12 @@ class SurfacePath implements ui.Path {
   /// a persistent div.
   ui.Rect? get webOnlyPathAsRect => pathRef.getRect();
 
+  /// Detects if path is a vertical or horizontal line returns LTRB or null.
+  ///
+  /// Used for web optimization of physical shape represented as
+  /// a persistent div.
+  ui.Rect? get webOnlyPathAsLine => pathRef.getLine();
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/path/path.dart:1544]: This doesn't need the `webOnly` prefix as it's in the private `_engine` library.

File: lib/web_ui/lib/src/engine/html/path/path_ref.dart [hunk 2]
```diff
@@ -173,9 +182,47 @@ class PathRef {
...
+      return null;
+    }
+    if ((x2 - x3) != width || (y3 - y0) != height) {
+      return null;
+    }
+    return ui.Rect.fromLTWH(x0, y0, width, height);
+  }
+
+  /// Returns horizontal/vertical line bounds or null if not a line.
+  ui.Rect? getLine() {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/path/path_ref.dart:213]: I'd give it a more specific name. When I read the call site I was immediately confused "What if the line is neither horizontal nor vertical?"

File: lib/web_ui/lib/src/engine/html/path/path_ref.dart [hunk 3]
```diff
@@ -173,9 +182,47 @@ class PathRef {
...
+    final double height = y2 - y1;
+    final double x3 = atPoint(3).dx;
+    final double y3 = atPoint(3).dy;
+    if (_fVerbs[3] != SPath.kLineVerb || y3 != y2) {
+      return null;
+    }
+    if ((x2 - x3) != width || (y3 - y0) != height) {
+      return null;
+    }
+    return ui.Rect.fromLTWH(x0, y0, width, height);
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/path/path_ref.dart:209]: Do we need to check that the path is closed?

File: lib/web_ui/lib/src/engine/html/path/path_ref.dart [hunk 4]
```diff
@@ -173,9 +182,47 @@ class PathRef {
...
-  ui.Rect _getRect() {
-    return ui.Rect.fromLTRB(
-        atPoint(0).dx, atPoint(0).dy, atPoint(1).dx, atPoint(2).dy);
+  ui.Rect? _detectRect() {
+    assert(_fVerbs[0] == SPath.kMoveVerb);
+    final double x0 = atPoint(0).dx;
+    final double y0 = atPoint(0).dy;
+    final double x1 = atPoint(1).dx;
+    final double y1 = atPoint(1).dy;
+    if (_fVerbs[1] != SPath.kLineVerb || y1 != y0) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/path/path_ref.dart:191]: Do we only detect rects that begin with a horizontal border?

File: lib/web_ui/lib/src/engine/html/path/path.dart [hunk 5]
```diff
@@ -1537,6 +1537,12 @@ class SurfacePath implements ui.Path {
   /// a persistent div.
   ui.Rect? get webOnlyPathAsRect => pathRef.getRect();
 
+  /// Detects if path is a vertical or horizontal line returns LTRB or null.
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/path/path.dart:1540]: nit: "... line. Returns ..."

**Review by yjbanov — APPROVED** [review]
lgtm

File: lib/web_ui/lib/src/engine/html/path/path.dart (hunk 5 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/path/path.dart]: Done.

File: lib/web_ui/lib/src/engine/html/path/path_ref.dart (hunk 4 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/path/path_ref.dart:193]: yes. added comment.

File: lib/web_ui/lib/src/engine/html/path/path_ref.dart (hunk 3 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/path/path_ref.dart:211]: Added warning to API, not to use for strokes only fills/clips.

File: lib/web_ui/lib/src/engine/html/path/path_ref.dart (hunk 2 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/path/path_ref.dart]: Done.

File: lib/web_ui/lib/src/engine/html/path/path.dart (hunk 1 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/path/path.dart]: Done.

**Review by yjbanov — APPROVED** [review]

**Pull request merged**